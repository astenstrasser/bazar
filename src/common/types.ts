export interface Product {
  _id: string;
  name: string;
  price: number;
  details?: string;
  imageUrl?: string;
  imageAltText?: string;
}
