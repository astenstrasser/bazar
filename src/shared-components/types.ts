export interface Product {
  _id: string;
  name: string;
  price: number;
  details?: string;
  imageUrl?: string;
  imageAltText?: string;
}

export enum Icon {
  LeftArrow,
  Login,
  Logout,
  Plus,
}
