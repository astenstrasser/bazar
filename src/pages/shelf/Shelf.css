.shelf {
  display: flex;
  justify-content: center;
  margin: auto;
  flex-wrap: wrap;
}

@media (max-width: 600px) {
  .shelf {
    width: 95%;
  }
}

@media (min-width: 600px) {
  .shelf {
    width: 90%;
  }
}

.create-icon {
  position: relative;
  bottom: 1rem;
  left: 1rem;
}
