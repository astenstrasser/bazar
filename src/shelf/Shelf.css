.shelf {
  display: flex;
  justify-content: center;
  margin: auto;
  flex-wrap: wrap;
}

@media (max-width: 600px) {
  .shelf {
    width: 95%;
  }
}

@media (min-width: 600px) {
  .shelf {
    width: 80%;
  }
}
