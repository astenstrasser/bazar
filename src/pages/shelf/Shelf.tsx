import React, { useEffect, useState } from "react";
import { useGoogleAuth } from "react-gapi-auth2";
import "./Shelf.css";
import ShelfItem from "./shelf-item/ShelfItem";
import { Icon, Product } from "../../shared-components/types";
import LoadingTag from "../../shared-components/loading-tag/LoadingTag";
import ActionIcon from "../../shared-components/action-icon/ActionIcon";
import { fetchAllProducts } from "../../api/api";

const Shelf: React.FC = () => {
  const [products, setProducts] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  const { googleAuth } = useGoogleAuth();

  useEffect(() => {
    isLoading &&
      fetchAllProducts()
        .then((res) => res.json())
        .then((products) => {
          setProducts(products);
          setIsLoading(false);
        });
    return () => setIsLoading(false);
  }, []);

  return (
    <>
      <div className="shelf">
        <LoadingTag isLoading={isLoading} />
        {products &&
          products.map((product: Product) => (
            <ShelfItem key={product._id} {...product} />
          ))}
      </div>
      {googleAuth?.isSignedIn.get() && (
        <ActionIcon
          onClick={() => {
            console.log("clicou");
          }}
          type={Icon.Plus}
          alt="Ícone para adicionar um produto"
        />
      )}
    </>
  );
};

export default Shelf;
