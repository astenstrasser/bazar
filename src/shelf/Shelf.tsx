import React, { useEffect, useState } from "react";
import "./Shelf.css";
import ShelfItem from "./shelf-item/ShelfItem";
import { Product } from "../common/types";

const Shelf: React.FC = () => {
  const [products, setProducts] = useState([]);

  useEffect(() => {
    const url = "http://bah-zar-api.herokuapp.com/product";
    fetch(url)
      .then((res) => res.json())
      .then((products) => {
        setProducts(products);
      });
  }, []);

  return (
    <div className="shelf">
      {products &&
        products.map((product: Product) => (
          <ShelfItem key={product._id} {...product} />
        ))}
    </div>
  );
};

export default Shelf;
