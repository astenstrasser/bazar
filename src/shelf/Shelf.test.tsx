import React from "react";
import { render, waitFor, screen } from "@testing-library/react";
import Shelf from "./Shelf";
import { rest } from "msw";

import { setupServer } from "msw/node";

import { Product } from "../common/types";
import { GApiProvider } from "react-gapi-auth2";

const mockResponse: Product[] = [
  {
    _id: "id1",
    name: "Mock Product 1",
    price: 100,
    imageUrl: "http://localhost:8080/image",
  },
  {
    _id: "id2",
    name: "Mock Product 2",
    price: 123,
    imageUrl: "http://localhost:8080/image2",
  },
];

const server = setupServer(
  rest.get("https://bah-zar-api.herokuapp.com/product", (req, res, ctx) => {
    return res(ctx.json(mockResponse));
  })
);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("renders snapshot", () => {
  const { container } = render(
    <GApiProvider clientConfig={{ client_id: "" }}>
      <Shelf />
    </GApiProvider>
  );
  expect(container).toMatchSnapshot();
});

test("renders data", async () => {
  const { container } = render(
    <GApiProvider clientConfig={{ client_id: "" }}>
      <Shelf />
    </GApiProvider>
  );

  await waitFor(() => screen.getAllByRole("heading"));

  expect(container).toMatchSnapshot();
});

test("should display loading tag while fetching data", async () => {
  server.use(
    rest.get("https://bah-zar-api.herokuapp.com/product", (req, res, ctx) => {
      setTimeout(() => {
        return res(ctx.json(mockResponse));
      }, 10);
    })
  );
  render(
    <GApiProvider clientConfig={{ client_id: "" }}>
      <Shelf />
    </GApiProvider>
  );

  expect(screen.getByText("Loading...")).toBeInTheDocument();
});
