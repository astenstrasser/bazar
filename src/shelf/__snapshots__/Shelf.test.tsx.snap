// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders data 1`] = `
<div>
  <div
    class="shelf"
  >
    <div />
    <div
      class="shelf-item"
      data-testid="shelf-item"
    >
      <div
        class="shelf-item--header"
      >
        <div
          class="shelf-item--info"
        >
          <h1>
            MOCK PRODUCT 1
             
          </h1>
          <p />
        </div>
        <div
          class="shelf-item--price"
        >
          R$ 
          100
        </div>
      </div>
      <img
        class="shelf-item--picture"
        src="http://localhost:8080/image"
      />
    </div>
    <div
      class="shelf-item"
      data-testid="shelf-item"
    >
      <div
        class="shelf-item--header"
      >
        <div
          class="shelf-item--info"
        >
          <h1>
            MOCK PRODUCT 2
             
          </h1>
          <p />
        </div>
        <div
          class="shelf-item--price"
        >
          R$ 
          123
        </div>
      </div>
      <img
        class="shelf-item--picture"
        src="http://localhost:8080/image2"
      />
    </div>
  </div>
</div>
`;

exports[`renders snapshot 1`] = `
<div>
  <div
    class="shelf"
  >
    <div>
      <span>
        Loading...
      </span>
    </div>
  </div>
</div>
`;
